### DISCUSSION TITLE
Ability to set mqtt v5.0 USER_PROPERTY dynamically via a code plugin

### DISCUSSION THREAD

**Discussion by jr1221** [nanomq/nanomq#1646] [General]
**Is your feature request related to a problem? Please describe.**
MQTT v5.0 has a USER_PROPERTY spec for a map of UTF 8 strings sent with payloads.  Right now, it seems nanomq only supports setting this via a conf file with static key/value string pairs.

**Describe the solution you'd like**
I would like a way to tell nanomq plugins to call a plugin that produces the key/value string pairs when each message must be sent, such like the [timestamp plugin in mosquitto](https://github.com/eclipse/mosquitto/blob/master/plugins/message-timestamp/mosquitto_message_timestamp.c).  My project's goal is to have timestamping be provided by the broker, but any way to generate the key/value pairs upon message creation would be helpful.

**Describe alternatives you've considered**
I have considered forking nanomq and adding it right into the code that sets the user_property statically, and will do so when I confirm that what I describe is not already in existence.

**Additional context**
Please let me know if this may already exist in some form here or in NanoNNG.  I know this is a big request but It would help nanomq surpass mosquitto.  Thanks for developing an awesome application!

**Comment by JaylinYu** [comment]
Really appreciate your suggestions. Currently plugin is not supported in nanomq. We have mobilized to add such a feature in next release.

It is also welcome If you wanna to implement this by yourself, just modify the msg in the state machine (broker.c)

**Comment by JaylinYu** [comment]
IT is done by #1606

**Comment by jr1221** [comment]
Works like a charm.  Thanks so much for the quick and effective feature yall added!  This program is really one of the fastest mqtt brokers out there!
Im gonna go ahead and close this...

**Comment by JaylinYu** [comment]
Thanks for your compliment. We will continue to work on that strange issue on aarch64. 

Would like to know more your project and use scenario. Is it related to https://electricracing.northeastern.edu？
Just curious about our role, knowing mre about community is the way to contribute more to open-source society.

**Comment by jr1221** [comment]
Hi, yes mqtt brokering is an important part of our Northeastern Electric Racing's telemetry framework.  A little background, we build formula-style electric racecars for competition around the U.S.  This is our 2nd car with a full telemetry system and the first using mqtt.  The way it works is multiple boards on the car generate CAN messages based on various states and sensors, and then a pi collects these messages and publishes it to nanomq as JSON.  Then nanomq (in its current planned state) bridges over QUIC on a long range 802.11ah network to a base station outside the car.  The base station nanomq or emqx broker then handles subscriptions for various telemetry fronted apps.  Additionally the in car dashboard subscribes ro what it needs to display to the driver.

The key point here is that the CAN data on the car is efficiently transmitted over a low bandwidth link so the pit crew and mechanical engineers can debug car performance real-time.  If you have any questions about this feel free!

**Comment by JaylinYu** [comment]
This is really impressive, seems like QUIC gonna play a crucial role in your case, will add QUIC listener make your solution more lightweight. 
I got some instant ideals:
Perhaps the QoS prioritization is also valuable to you. let critical CAN-ID messages pass first.

Do you need some sort of edge computing solution to filter CAN data? like time-triggering, always maintaining a high volume of transmission looks like a point to improve efficiency. But I believe this is already be done by your program.

you want timestamp in msg, I guess you are using TSN, do you require ACF/AVTP format data conversion? Is DDS an interesting thing to you?

**Comment by jr1221** [comment]
Thanks for the QUIC interest.  Yes the main goal of our technology stack is to increase the reliability of data transfer in unstable networks.  Our 900 mhz link has an average throughput of 1.3 mb/s but there are scenarios at which 200 kb/s is the best we can do, as the racetrack is long distance.  We would love any tips you have on improving QUIC/nanomq reliability in such an environment.

Also ,the head of our data & controls team specifcally wanted to ask if there was a way to diff message states over the quic bridge such that identical messages are cached on each end of the bridge rather than being transmitted through the bridge, possibly via broadcasting deltas.

In regards to both QoS and CAN filtering, We also are planning to universally have QoS zero messages, and are not worried about priorities or real-time functions as our telemetry system has no real control sync role-- it only records debug information.  All functional changes external to the CAN network are made real-time are by the car's human driver, and all internal communications run embedded firmware, usually RTOS systems.  I think this is a reason DDS may not be needed by our team, as we only really have one true linux system in operation.  However I may be misunderstanding the use-case of DDS.


In terms of timestamping I am not familiar with the data conversions you are describing.  We simply have a universal source of time via the car pi, and all messages are timestamped with simple unix time when they reach the broker.  This time isn;t very accurate to when they were sent via CAN bus, however it is comparative enough for our non-essential usecase.

**Comment by JaylinYu** [comment]
Just release the 0.21.1, along with several fixes and new features. Plz update to the new version. @jr1221 

Even without priority in your packets, the UDP + multi-stream still beats TCP in wireless mode, especially when you transmit 1.3mb/s over a 900 MHz radio. This is how congestion happens, gonna cause TCP disconnects.

As for DDS, there is no good implementation of DDS on RTOS yet, and it is definitely not needed if you are running a one-to-many telemetry-only network.

Would like to know more about " diff message states".  It sounds like a data synchronization of network disk to me, just similar to NAS. What goal do you want to accomplish? Perhaps we can make it happen in the newly release feature-- MQTT Stream.

**Comment by jr1221** [comment]
I just upgraded the system to 0.21.1 and it works great.  So does the timestamping functionality.  We will be testing our full network in the next few weeks and are excited to see the performance of nanomq under load.  Our rather unscientific benchmarks show it is bound to work great!

In terms of the message diffing idea:
It was more a theoretical idea of looking for when a message changes in order to save bandwidth.  Since our telemetry system is very verbose, there is often republishing of an identical message.  We could implement a sort of cache in each publisher and subscriber however that has a complexity overhead and does not indicate whether a publisher goes offline.

We were wondering under the scenario that there is a repetitive publishing of a message to a specific topic, upon recieving that message the broker could decide if that message has changed when compared to the last message of the topic. If it has not changed then the broker could send a "notification" over the bridge.  The other member of the bridge can then produce to its subscribers the same cached message. 
 This "notification" of sorts could help to save bandwidth across the bridge.

The more I describe it the more I realize how such a feature would have to be a rather ingrained low level feature, of which there probably is none.

**Reply by JaylinYu** [reply]
Sure, your proposal is one of our initiatives to start nanomq+ekuiper combo. check https://ekuiper.org/.
You can setup a stream and compare the difference each time.
And there is a more smart way to diff the msg/data-space, using a remote synchronization methodology, which is another feature planned in nanomq.
